langkit/support/langkit_support-diagnostics-output.adb

```diff
@@ -0,0 +1,181 @@
+with Ada.Strings.Wide_Wide_Fixed; use Ada.Strings.Wide_Wide_Fixed;
+with Ada.Text_IO; use Ada.Text_IO;
+with Ada.Wide_Wide_Text_IO; use Ada.Wide_Wide_Text_IO;
+
+with GNATCOLL.Terminal; use GNATCOLL.Terminal;
+
+with Langkit_Support.Images; use Langkit_Support.Images;
+
+package body Langkit_Support.Diagnostics.Output is
+
+   Term_Info : Terminal_Info;
+   Colors_Init : Boolean := False;
+
+   use Ada.Containers;
+
+   procedure Print_Source_Listing
+     (Sloc        : Source_Location_Range;
+      Buffer      : in out Text_Buffer;
+      Lines_After : Natural := 0);
+   --  Print a source listing
+
+   procedure Reset_Colors;
+   --  Reset the state of colors for ``Term_Info``
+
+   ------------------
+   -- Reset_Colors --
+   ------------------
+
+   procedure Reset_Colors is
+   begin
+      Set_Color (Term_Info, Foreground => Reset);
+      Set_Style (Term_Info, Reset_All);
+   end Reset_Colors;
+
+   --------------
+   -- Get_Line --
+   --------------
+
+   function Get_Line
+     (Self        : in out Text_Buffer;
+      Line_Number : Positive) return Text_Type
+   is
+      T      : Text_Type renames Self.Text_Ptr.all;
+      Offset : Natural := 0;
+   begin
+      --  If there are line spans, then the start offset is the offset of the
+      --  last computed line span.
+      if Self.Line_Spans.Length > 0 then
+         Offset := Self.Line_Spans.Last_Element;
+      end if;
+
+      --  Compute every line span between the last computed line span and the
+      --  queried one.
+      for I in Positive (Self.Line_Spans.Length + 1) .. Line_Number loop
+         Offset := Index (T, Chars.LF & "", Offset + 1);
+
+         --  If no next newline found, put the next offset just after the end
+         --  of the string, at the (virtual) "$" EOF character.
+         if Offset = 0 then
+            Offset := T'Last + 1;
+         end if;
+         Self.Line_Spans.Append (Offset);
+      end loop;
+
+      return  T ((if Line_Number = 1
+                 then 0
+                 else Self.Line_Spans (Line_Number - 1)) + 1
+                 .. Self.Line_Spans (Line_Number) - 1);
+
+   end Get_Line;
+
+   --------------------------
+   -- Print_Source_Listing --
+   --------------------------
+
+   procedure Print_Source_Listing
+     (Sloc        : Source_Location_Range;
+      Buffer      : in out Text_Buffer;
+      Lines_After : Natural := 0)
+   is
+      procedure Start_Line (Line_Nb : String := "");
+
+      Line_Nb : constant Positive := Positive (Sloc.Start_Line);
+      Start_Offset : constant Positive := Positive (Sloc.Start_Column);
+      End_Offset   : constant Positive := Positive (Sloc.End_Column);
+
+      Line_Nb_Width : constant Positive :=
+        Positive'Image (Line_Nb + Lines_After)'Length - 1;
```

Comment:
It looks like this procedure assumes that both the start and the end of `Sloc` (which should be called `Sloc_Range`, right?) are on the same line. Could you clarify if that should a precondition, or how it should be handled?

Reply:
Ok, I just saw the comment below ;-) Could you move it to the spec?